rust/src/smb/detect.rs

```diff
@@ -201,3 +206,435 @@ pub extern "C" fn rs_smb_tx_get_dce_iface(state: &mut SMBState,
     }
     return 0;
 }
+
+
+#[no_mangle]
+pub unsafe extern "C" fn rs_smb_cmd_match(
+    tx: &mut SMBTransaction, cmd_valid_codes: &mut SmbCmdValidCodes,
+) -> u8 {
+
+    let version = tx.vercmd.get_version();
+    let cmd;
+    if version == 1 {
+        cmd = tx.vercmd.get_smb1_cmd().1.into();
+    } else {
+        cmd = tx.vercmd.get_smb2_cmd().1;
+    }
+
+    SCLogDebug!("rs_smb_cmd_match: version {} cmd {}", version, cmd);
+
+    if let Some(valid_codes) = cmd_valid_codes.0.get(&version) {
+        if valid_codes.contains(&cmd) {
+            return 1;
+        }
+    }
+
+    return 0;
+}
+
+
+#[no_mangle]
+pub unsafe extern "C" fn rs_smb_cmd_parse(carg: *const c_char) -> *mut c_void {
+    if carg.is_null() {
+        return std::ptr::null_mut();
+    }
+    let arg = match CStr::from_ptr(carg).to_str() {
+        Ok(arg) => arg,
+        _ => {
+            return std::ptr::null_mut();
+        }
+    };
+
+    match parse_cmd_data(arg) {
+        Ok(detect) => Box::into_raw(Box::new(detect)) as *mut _,
+        Err(_) => std::ptr::null_mut(),
+    }
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn rs_smb_cmd_free(ptr: *mut c_void) {
+    if ptr != std::ptr::null_mut() {
+        std::mem::drop(Box::from_raw(ptr as *mut SmbCmdValidCodes));
+    }
+}
+
+/// Stores the SMB command codes used to match with smb.cmd keyword.
+/// It includes both valid codes for SMB1 and SMB2.
+#[derive(Debug, PartialEq)]
+pub struct SmbCmdValidCodes (HashMap<u8, HashSet<u16>>);
```

Comment:
`HashMap<u8, HashSet<u16>>` Is the u8 the version being 1 or 2 ?
In this case, do we need a HashMap instead of 2 fields ?

Reply:
Sure, that could be a better idea. Implemented in https://github.com/OISF/suricata/pull/7181/commits/d3da2fdbb6043b087e4d9abc4ff769c67765c05e.